be/src/exec/pipeline/sink/connector_sink_operator.cpp

```diff
@@ -0,0 +1,162 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "connector_sink_operator.h"
+
+#include "formats/utils.h"
+#include "glog/logging.h"
+
+namespace starrocks::pipeline {
+
+ConnectorSinkOperator::ConnectorSinkOperator(OperatorFactory* factory, const int32_t id, const int32_t plan_node_id,
+                                             const int32_t driver_sequence,
+                                             std::unique_ptr<connector::ConnectorChunkSink> connector_chunk_sink,
+                                             FragmentContext* fragment_context)
+        : Operator(factory, id, "connector_sink_operator", plan_node_id, false, driver_sequence),
+          _connector_chunk_sink(std::move(connector_chunk_sink)),
+          _fragment_context(fragment_context) {}
+
+Status ConnectorSinkOperator::prepare(RuntimeState* state) {
+#ifndef BE_TEST
+    RETURN_IF_ERROR(Operator::prepare(state));
+#endif
+    RETURN_IF_ERROR(_connector_chunk_sink->init());
+    return Status::OK();
+}
+
+void ConnectorSinkOperator::close(RuntimeState* state) {
+    if (_is_cancelled) {
+        while (!_rollback_actions.empty()) {
+            _rollback_actions.front()();
+            _rollback_actions.pop();
+        }
+    }
+#ifndef BE_TEST
+    Operator::close(state);
+#endif
+}
+
+bool ConnectorSinkOperator::need_input() const {
+    if (_no_more_input) {
+        return false;
+    }
+
+    while (!_add_chunk_future_queue.empty()) {
+        // cannot accept chunk if any add_chunk_futures is not ready
+        if (!is_ready(_add_chunk_future_queue.front())) {
+            return false;
+        }
+        if (auto st = _add_chunk_future_queue.front().get(); !st.ok()) {
+            LOG(WARNING) << "cancel fragment: " << st;
+            _fragment_context->cancel(st);
```

Comment:
Will this operation be a time-cost operation? or block operation?

Reply:
No. It is neither time-cost nor blocking.